### DISCUSSION TITLE
Code TODO List

### DISCUSSION THREAD

**Discussion by scottfurry** [sqlitebrowser/sqlitebrowser#3419] [Ideas]
### Background
Walking through code progressing settings handling update, I'm struck by a spat of changes _that could be made_ to clean things up. However, these changes would not be apart of updating settings. I won't try to _slip_ changes into the mix that have nothing to do with the problem at hand. However, I also do not want to lose visibility of these changes.

When in doubt...make a *_TODO_* list

If you feel inclined, you're welcome to grab an item and make it _go away_ with a pull request. Please let me know here once completed in case I miss the conversation thread elsewhere.

#### Renaming Menu Items Removing Ampersands
Ref: `MainWindow.cpp:504`
Is this still a thing in Qt 5.15 | Qt.6.X? Multiplatform checking necessary.

#### Consistent UI Item Naming Practice
Ref: `MainWindow.cpp: 2354, 2356`
We have `dbToolBar` as a ui element name followed by a similar item named `toolbarSql`. Go through the source code and ensure ui elements have a coherent and non-clashing naming scheme that is consistently applied throughout.

#### Consistent Pointer|Reference Usage
Ref: Multiple Files
Ensure code is following a clear and consistent pattern when using references and pointers. This is pretty much the code version of _nails-on-chalkboard_ for me.  I see both variations of...
`<type>[*|&][space][variable]`
and
`<type>[space][*|&][variable]`

Pick *ONE* and ensure that it is enforced throughout all header and source files.
###### sidebar
I tried to pull in [astyle](https://astyle.sourceforge.net/) a couple years back. It was preferable than [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) as not all users/platforms could easily incorporate llvm. Pointer|Reference type usage is *one aspect of many* that `astyle` could address. There are other code formatters that could be employed. When _set up_ they work beautifully. Getting there is the challenge.

#### License Headers
Ref: Multiple Files
I was surprised to edit code to find that multiple source files are missing a licensing attribute at the top. Yup, it is repetitive, noisy legalese. A necessary evil in our modern, digital age. Ensure others know to respect distribution requirements.

See [this](https://www.gnu.org/licenses/gpl-howto.html) about how to use the GPL or [this](https://www.mend.io/blog/top-10-gpl-license-questions-answered/) for common GPL questions.

#### Expand Command Line Abilities
Ref #3422
Command Line flags of DB4S should be expanded to include select flags of `sqlcipher` command line. Specifically, the `sqlcipher` flag `init` should be incorporated to allow decryption of a database from command line.

Work may involve incorporating a better software library to better help manage and maintain command line flags.

#### Verify Switch Statements with Case Structure Returns
Ref: `DbStructureModel.cpp:36` and others

Function `QVariant DbStructureModel::data(const QModelIndex& index, int role) const` returns a QVariant from within a case statement in a switch structure. However, there are **NO** `break` statements prior to return.

Also, there was a recent discussion of an issue (Issue # TBD) where the fix was to rework the return from the switch structure. Unfortunately, the resolution caused another compiler warning about the switch structure and function return.

Verify these types of switch structures in the source code in order to silence any/all compiler/static analysis warnings.

There is also an issue in switch structure formatting. The structures are inconsistently formatted. This has a similar solution as work item **Renaming Menu Items Removing Ampersands** above. Automated Code Formatting tools would correct this problem easily but would require identification/setup of a format to be applied.

#### Remove `sourceforge` domain in project files
Ref: [comment to #3341](https://github.com/sqlitebrowser/sqlitebrowser/issues/3341#issuecomment-1541090992)

Reference identified usage of `sourceforge` domain for storage of settings. Matter is specific to MacOS where a plist file explicitly uses URL pointing to DB4S project on `sourceforge`. Item was thought to be originally addressable via work under #3317. Search of sources found numerous references to reverse URL. Removal of `sourceforge` domain requires more involved effort that needs to addressed independently.

```
$ ack sourceforge
src/app.plist
62:	<string>net.sourceforge.sqlitebrowser</string>

README.md
249:* https://sourceforge.net/projects/sqlitebrowser

installer/macos/build_sqlitebrowser_nightly-intel.sh
176:REQUEST_UUID=`xcrun altool --notarize-app --primary-bundle-id "net.sourceforge.sqlitebrowser" -u "${APPLE_ID}" -p @env:NOTARIZATION_PASSWORD --output-format json --file DB\ Browser\ for\ SQLite-intel_${DATE}.dmg 2>&1 | jq -r '."notarization-upload"."RequestUUID"'`
313:REQUEST_UUID=`xcrun altool --notarize-app --primary-bundle-id "net.sourceforge.sqlitebrowser" -u "${APPLE_ID}" -p @env:NOTARIZATION_PASSWORD --output-format json --file DB\ Browser\ for\ SQLCipher-intel_${DATE}.dmg 2>&1 | jq -r '."notarization-upload"."RequestUUID"'`

```

#### Update Dependent Libraries and Document Update Methodology
Ref: `[project root]/libs/` directory

Current project dependencies, QScintilla et al, are lagging behind. People are busy. Others are not able to pick up the slack because of a lack of knowledge or ability. Personally, I tripped over a possible updating methodology for QScintilla as part of #3410.

###### QScintilla
Internal version 2.11.1 from 2019.
Latest upstream is 2.14.1.

###### QHexEdit
Internal version 0.8.6 from 2019.
Latest upstream is 0.8.9 from 2020.
**NOTE:** we should have a mitigation strategy if this dependency is not being actively maintained.

###### QCustomPlot
Internal version 2.1.0 from March 2021.
Latest upstream is version 2.1.1 from November, 2022.

###### nlohmann/json
Internal version 3.10.4 from October, 2021.
Latest upstream is version 3.11.2 dated August, 2022.

###### ColinDuquesnoy/QDarkStyleSheet
Internal version is unknown and not documented
Latest upstream is version 3.0.2 dated April, 2021.

The key takeaway on this item is **Update and Document how** for the next person.

#### Update Preferences Dialog - Correct Poor Font Size Display
ref: image
![showing_pref_font_size_squeeze](https://github.com/sqlitebrowser/sqlitebrowser/assets/6686571/05d709ed-c680-4c42-ac53-c2365c7492fe)

Under Qt 5.15.10 on Linux, control sizing/layout needs to be improved to prevent poor presentation of font size. Image above shows example of current experience (2023Aug). Suspect `sizeHint()` code needs tweaking. Problem is persistent and occurs for both Light and Dark Mode presentations.

#### Update CMake Usage - Deprecation - Minimum Version
ref: [CMake 3.27 Release Notes - Deprecation](https://cmake.org/cmake/help/latest/release/3.27.html#deprecated-and-removed-features)

Recent build on MacOS with latest CMake installed resulted in numerous warnings. Issue stems from Kitware implementing a deprecation warning for calling out `cmake_minimum_required` or `cmake_policy` with a version **LESS THAN** 3.5. Multiple project `cmake` files still use a minimum CMake of 2.8.X.

_more as required_
...

**Comment by justinclift** [comment]
Cool!  Really, really good approach. :smile:

**Comment by lucydodo** [comment]
Why not use [SPDX notation](https://spdx.dev/) for licenses in each source and header file?

**Reply by scottfurry** [reply]
Because I didn't know it existed? There's some reading here.

My first pass of reading indicates there's a standard behind this. Great, but we still need to identify a license for code. My understanding is the SPDX is not a license but a standard for formatting and communicating licensing information.

**Comment by scottfurry** [comment]
Appending work item **Expand Command Line Abilities** to list

**Comment by scottfurry** [comment]
Appending work item **Verify Switch Statements with Case Structure Returns**

**Reply by justinclift** [reply]
That sounds like something that existing analysis tools should already be doing?

**Reply by scottfurry** [reply]
_Should be doing_. If only we were implementing a standard for the project?

This is where the *sidebar* discussion under the **Consistent Pointer|Reference Usage** was leading. As noted, once a code formatter is identified, a project standard defined, and everything setup - a multitude of _code sins_ go away. The formatter takes care of these things _automagically_.

The challenge is to say "this is the standard for the project" and be consistent in it's application.

**Comment by scottfurry** [comment]
Appending work item **Remove `sourceforge` domain in project files**

**Comment by scottfurry** [comment]
Appending work item **Update Dependent Libraries and Document Update Methodology**

**Reply by scottfurry** [reply]
Note: Deprecation warning in QHexEdit during build.
```
[6/129] Building CXX object libs/qhexedit/CMakeFiles/qhexedit.dir/src/qhexedit.cpp.o
{project_root}/libs/qhexedit/src/qhexedit.cpp: In member function ‘void QHexEdit::setFont(const QFont&)’:
{project_root}/libs/qhexedit/src/qhexedit.cpp:435:39: warning: ‘int QFontMetrics::width(QChar) const’ is deprecated: Use QFontMetrics::horizontalAdvance [-Wdeprecated-declarations]
  435 |     _pxCharWidth = fontMetrics().width(QLatin1Char('2'));
      |                    ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
In file included from /usr/include/qt5/QtGui/qpainter.h:59,
                 from /usr/include/qt5/QtGui/QPainter:1,
                 from {project_root}/libs/qhexedit/src/qhexedit.cpp:4:
/usr/include/qt5/QtGui/qfontmetrics.h:110:9: note: declared here
  110 |     int width(QChar) const;
      |         ^~~~~
```
Situation is corrected upstream in latest release.

**Reply by scottfurry** [reply]
Expanding work item to include **ColinDuquesnoy/QDarkStyleSheet**
Internal version in project is unknown and not documented.

**Comment by lucydodo** [comment]
@scottfurry Do you mind if I put some of own to-dos in here, and track them?
(I'm **NOT** tryinjg to give anyone a to-do, I just want to discuss them together :smile: )

**Reply by scottfurry** [reply]
@lucydodo - meet me on Gitter - we can discuss some concerns there.

**Reply by lucydodo** [reply]
Well, Gitter is good. but I have a few to-dos for project that I've been thinkg about,
and I thought it would be be nice to write them down here so I can discuss them with others. 😄 
(Actually, I think this discussion article can set the stage for the next release)

**Reply by scottfurry** [reply]
Gitter, please. We can compare notes.

**Comment by scottfurry** [comment]
Appending work item **Update Preferences Dialog - Correct Poor Font Size Display**

**Reply by scottfurry** [reply]
Problem doesn't occur under "Follow Desktop Style". Issue may be isolated to theme/styling.

**Comment by scottfurry** [comment]
Appending work item **Update CMake Usage - Deprecation - Minimum Version**

**Comment by justinclift** [comment]
Re-opening this issue, as it's still actively used. :smile:
